File: packages/spoke-operations/src/tools/orchestrate.ts
/**
 * Orchestrated remote tool invocation over an injected `ToolInvokePort`.
 *
 * Frozen sequence (`tool-contracts.md` §5; no `requireToolInvokePort`
 * wrapper): (1) `parseToolCapabilityId(request.capability_id)` grammar gate
 * → `INVALID_INPUT`; (2) runtime guard `typeof port.invokeTool ===
 * "function"` → else `CAPABILITY_PORT_MISSING` with
 * `details.capability = request.capability_id` (per-tool blame — observable
 * at the JS boundary where structural typing cannot protect dynamic
 * composition: plain objects reject structurally instead of throwing a
 * `TypeError`); (3) `port.invokeTool(request)` returned as-is.
 *
 * Argument validation is NOT re-run here — callers run the structural
 * argument gate (`validateToolArguments`, see `tools/helpers.ts`) before
 * orchestrating; this function validates only request grammar.
 */

import {
  SpokeRejectCode,
  spokeReject,
  type SpokeResult,
} from "../result.js";
import { parseToolCapabilityId } from "./helpers.js";
import type {
  ToolInvokePort,
  ToolInvokeRequest,
  ToolInvokeResponse,
} from "./port.js";

export async function orchestrateInvokeTool(
  port: ToolInvokePort,
  request: ToolInvokeRequest,
): Promise<SpokeResult<ToolInvokeResponse>> {
  const grammar = parseToolCapabilityId(request.capability_id);
  if (!grammar.ok) {
    return grammar;
  }
  if (typeof port.invokeTool !== "function") {
    return spokeReject(
      SpokeRejectCode.CAPABILITY_PORT_MISSING,
      `Missing ToolInvokePort for capability "${request.capability_id}" (invokeTool is not a function)`,
      { capability: request.capability_id },
    );
  }
  return port.invokeTool(request);
}

File: crates/spoke-operations/src/tools/orchestrate.rs
//! Orchestrated remote tool invocation over an injected `ToolInvokePort`
//! (parity with `packages/spoke-operations/src/tools/orchestrate.ts`).
//!
//! Frozen sequence (`tool-contracts.md` §5; no `requireToolInvokePort`
//! wrapper): (1) `parse_tool_capability_id(request.capability_id)` grammar
//! gate → `INVALID_INPUT`; (2) the port is `&dyn ToolInvokePort`, so absence
//! is unrepresentable at the type level — the negative test pins a
//! `MissingToolInvokePort` double whose `invoke_tool` returns the same
//! `CAPABILITY_PORT_MISSING` reject with `details.capability`, demonstrating
//! code parity with the TS runtime guard; (3) `port.invoke_tool(request)`
//! returned as-is.
//!
//! Argument validation is NOT re-run here — callers run the structural
//! argument gate ([`crate::validate_tool_arguments`]) before orchestrating;
//! this function validates only request grammar.

use crate::result::SpokeResult;
use crate::tools::helpers::parse_tool_capability_id;
use crate::tools::port::{ToolInvokePort, ToolInvokeRequest, ToolInvokeResponse};

pub async fn orchestrate_invoke_tool(
    port: &dyn ToolInvokePort,
    request: ToolInvokeRequest,
) -> SpokeResult<ToolInvokeResponse> {
    match parse_tool_capability_id(&request.capability_id) {
        SpokeResult::Reject(reject) => SpokeResult::Reject(reject),
        SpokeResult::Ok(_) => port.invoke_tool(request).await,
    }
}
